File: bin/satest/bench/1.js
function runTests() {
  // --- TEST u64 ADD ---
  let r1_u64 = 100n;
  let r2_u64 = 200n;
  let r3_u64 = r1_u64 + r2_u64;

  // --- TEST u32, u16, 2xu8 ADD ---
  // Packing: [ u32 | u16 | u8 | u8 ]
  const r4 = (200n << 32n) | (30n << 16n) | (200n << 8n) | 30n;
  const r5 = (800n << 32n) | (900n << 16n) | (56n << 8n) | 30n;

  // u32 ADD (+ offset 1)
  let res_u32 = ((r4 >> 32n) + (r5 >> 32n) + 1n) & 0xffffffffn;

  // u16 ADD (+ offset 1)
  let res_u16 =
    (((r4 >> 16n) & 0xffffn) + ((r5 >> 16n) & 0xffffn) + 1n) & 0xffffn;

  // SIMD u8 ADD (+ offset 0)
  let res_u8_a = (((r4 >> 8n) & 0xffn) + ((r5 >> 8n) & 0xffn)) & 0xffn;
  let res_u8_b = ((r4 & 0xffn) + (r5 & 0xffn)) & 0xffn;

  // Re-packing r6
  return (res_u32 << 32n) | (res_u16 << 16n) | (res_u8_a << 8n) | res_u8_b;
}

const iterations = 10000;
const latencies = new BigUint64Array(iterations);

// Warm up - Let V8 optimize the function
for (let i = 0; i < 2000; i++) runTests();

for (let i = 0; i < iterations; i++) {
  const start = process.hrtime.bigint();
  runTests();
  const end = process.hrtime.bigint();
  latencies[i] = end - start;
}

// Calculate Median
latencies.sort();
const median = latencies[iterations / 2];
const result = runTests();

console.log(`Final Packed Register (r6): 0x${result.toString(16)}`);
console.log(`Median Execution Time: ${median} ns`);

File: bin/satest/bench/1.py
import time
import statistics

def run_tests():
    # --- TEST u64 ADD ---
    # r1 = 100, r2 = 200
    r1_u64 = 100
    r2_u64 = 200
    r3_u64 = r1_u64 + r2_u64 # Result: 300

    # --- TEST u32, u16, 2xu8 ADD ---
    # Packing: [ u32 | u16 | u8 | u8 ]
    def pack_val(v32, v16, v8_a, v8_b):
        return (v32 << 32) | (v16 << 16) | (v8_a << 8) | v8_b

    def unpack_val(val):
        return {
            "u32": (val >> 32) & 0xFFFFFFFF,
            "u16": (val >> 16) & 0xFFFF,
            "u8_a": (val >> 8) & 0xFF,
            "u8_b": val & 0xFF
        }

    r4 = pack_val(200, 30, 200, 30)
    r5 = pack_val(800, 900, 56, 30)

    # u32 ADD (+ offset 1)
    # The vadd offset 1 suggests adding 1 to the result of the packed components
    res_u32 = ((r4 >> 32) + (r5 >> 32) + 1) & 0xFFFFFFFF
    
    # u16 ADD (+ offset 1)
    res_u16 = (((r4 >> 16) & 0xFFFF) + ((r5 >> 16) & 0xFFFF) + 1) & 0xFFFF
    
    # SIMD u8 ADD (+ offset 0)
    # Note: SIMD u8 adds both u8 slots independently
    res_u8_a = (((r4 >> 8) & 0xFF) + ((r5 >> 8) & 0xFF)) & 0xFF
    res_u8_b = ((r4 & 0xFF) + (r5 & 0xFF)) & 0xFF

    # Re-packing r6 (simulating the final state)
    r6 = (res_u32 << 32) | (res_u16 << 16) | (res_u8_a << 8) | res_u8_b
    return r6

# Execution loop
latencies = []

for _ in range(10_000):
    start = time.perf_counter_ns()
    run_tests()
    end = time.perf_counter_ns()
    latencies.append(end - start)

median_latency = statistics.median(latencies)

print(f"Median Execution Time: {median_latency:.2f} ns")